desc/compute/utils.py

```diff
@@ -0,0 +1,303 @@
+        )
+        zeta_surface = (
+            jnp.repeat(x, repeats=nodes_per_theta_curve)
+            if no_jax
+            else jnp.repeat(
+                x,
+                repeats=nodes_per_theta_curve,
+                total_repeat_length=nodes_per_zeta_surface,
+            )
+        )
+        return jnp.tile(zeta_surface, reps=grid.num_zeta)
+
+    if surface_label == "theta":
+        assert len(x) == grid.num_theta
+        if isinstance(grid, ConcentricGrid):
+            raise NotImplementedError(
+                "expand(ConcentricGrid, surface_label='theta') not implemented yet"
+            )
+
+        nodes_per_theta_surface = grid.num_nodes // grid.num_theta
+        assert grid.num_nodes == grid.num_theta * nodes_per_theta_surface
+        theta_sort_x = (
+            jnp.repeat(x, repeats=nodes_per_theta_surface)
+            if no_jax
+            else jnp.repeat(
+                x, repeats=nodes_per_theta_surface, total_repeat_length=grid.num_nodes
+            )
+        )
+        # arrange theta sorted x to match grid.nodes ordering
+        theta_sort_inverse_idx = jnp.argsort(jnp.argsort(grid.nodes[:, 1]))
+        return theta_sort_x[theta_sort_inverse_idx]
+
+    if surface_label == "zeta":
+        assert len(x) == grid.num_zeta
+        nodes_per_zeta_surface = grid.num_nodes // grid.num_zeta
+        if no_jax:
+            return jnp.repeat(x, repeats=nodes_per_zeta_surface)
+        return jnp.repeat(
+            x, repeats=nodes_per_zeta_surface, total_repeat_length=grid.num_nodes
+        )
+
+    raise ValueError("Surface label must be 'rho', 'theta', or 'zeta'.")
+
+
+def surface_integrals(grid, q=1, surface_label="rho", match_grid=False):
+    """
+    Bulk surface integral function.
+    Computes the surface integral of the specified quantity for all surfaces in the grid.
+
+    Parameters
+    ----------
+    grid : Grid, LinearGrid, ConcentricGrid, QuadratureGrid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to integrate.
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute integration over.
+        Note for surface_label="theta" LinearGrid will have better accuracy than QuadratureGrid.
+    match_grid : bool
+        False to return a compressed array which assigns every surface integral to a single element.
+        This array will have length = number of unique surfaces in the grid.
+        True to return an expanded array which assigns every surface integral to all indices in
+        grid.nodes which are associated with that surface.
+        This array will match the grid's pattern and have length = grid.num_nodes.
+
+    Returns
+    -------
+    integrals : ndarray
+        Surface integrals of q over each surface in grid.
+    """
+    if surface_label == "theta" and not isinstance(grid, LinearGrid):
+        print("Nonlinear grids may have bad accuracy for theta surface computations.")
```

Comment:
why would a nonlinear grid be that much worse?
Also this might be better as a warning rather than regular print

Reply:
I think it's because QuadratureGrid does not have nodes linearly spaced in rho.
```
%matplotlib inline
from desc.plotting import plot_grid
from desc.grid import *
from desc.compute.utils import surface_integrals
import numpy as np

g1 = LinearGrid(L=2, M=0, N=0)
g2 = QuadratureGrid(L=1, M=0, N=0)
plot_grid(g1)
plot_grid(g2)

# We want both of these to equal the area of a theta surface = 4pi^2 (39.4...)
# But non-linear grids are less accurate.
# The relevant difference between the two grids is that LinearGrid has nodes linearly spaced in rho.
# As L approached high values, the area approaches 4pi^2 on quadrature grids too.
area_g1 = surface_integrals(g1, surface_label="theta")
area_g2 = surface_integrals(g2, surface_label="theta")
print(area_g1)
print(area_g2)
# This is what surface_integrals will return for theta, M=0.
assert g1.M != 0 or np.allclose(area_g1, g1.spacing[:, [0, 2]].prod(axis=1).sum())
assert g2.M != 0 or np.allclose(area_g2, g2.spacing[:, [0, 2]].prod(axis=1).sum())

# Should the spacing sum to 4 pi^2 on QuadratureGrids? If yes this is also a bug on master.
print(g1.spacing.prod(axis=1).sum())
print(g2.spacing.prod(axis=1).sum())
```

Reply:
ahhh yeah i forgot the quadrature weights dont always sum to 4pi^2. 
I'd still change this to a warning, otherwise fine

Reply:
Thanks for fixing this for QuadratureGrids Daniel! Only thing is ConcentricGrids have the same issue. If the rho spacing isn't going to be changed for that, we should narrow the warning to concentric grids.

Reply:
The rho spacing is correct for the `ConcentricGrid` as well, but it only really makes sense on the theta=0 surface: 

```import numpy as np
from desc.plotting import plot_grid
from desc.grid import LinearGrid, QuadratureGrid, ConcentricGrid
from desc.compute.utils import surface_integrals

g1 = LinearGrid(L=2, M=0, N=0)
g2 = QuadratureGrid(L=2, M=0, N=0)
g3 = ConcentricGrid(L=4, M=4, N=0)

plot_grid(g1)
plot_grid(g2)
plot_grid(g3)

# theta surface integrals should be 2*pi = 6.283
print(surface_integrals(g1, surface_label="theta")[0])
print(surface_integrals(g2, surface_label="theta")[0])
print(surface_integrals(g3, surface_label="theta")[0])

# grid spacing should sum to 4*pi^2 = 39.478
print(g1.spacing.prod(axis=1).sum())
print(g2.spacing.prod(axis=1).sum())
print(g3.spacing.prod(axis=1).sum())
```

It would be a very unusual use case to evaluate a theta surface integral on a `ConcentricGrid`, because basically every node has a unique theta coordinate.

Reply:
Actually, I'm not sure if my change to the `QuadratureGrid` was good or not. With my edit to the drho spacing, the integral over a theta surface does correctly evaluate to 2*pi (as shown above). But it also breaks the cross-section area computation given by `data["A"]`, which is very bad. 

So perhaps the `QuadratureGrid` does need to keep both its special spacing & weights. But something still feels strange.

Reply:
- Quad grid before divide by r sums to 1/2 but is expected as the weight fxn is r an domain is from 0 to 1